### PR TITLE
Transformers series part 2

### PR DESCRIPTION
**Pull request by iulusoy**
Transformers series part 2, please continue to ignore the notebook.
We transform the pre-tokenized data from spacy, that is now stored in a nested list of lists - each of these lists contains the tokens (labels) for each sentence - into a further tokenized list, where the words are split into pieces. The new splitting of the words needs to be accounted for in the labels, thus a map is applied to transform the labels in accordance with the splitting of the pre-tokenized tokens.

- tokenize pre-tokenized data with the selected tokenizer
- extend labels for new tokens
- apply map from pre-tokenized to tokenized data for subsequent processing through the model

### CODE REVIEW

File: moralization/transformers_data_handler.py [hunk 1]
```diff
@@ -94,3 +101,67 @@ def init_tokenizer(self, model_name="bert-base-cased", kwargs=None):
             raise ValueError(
                 "Please use a different model that provices a fast tokenizer"
             )
+
+    def tokenize(self, wordlist: list = None):
```
Comment (lkeegan) [inline moralization/transformers_data_handler.py:105]: If we want to support python 3.7 I think we have to use `List` for type hints

`from typing import List`

`def tokenize(self, wordlist: List = None):`

File: moralization/transformers_data_handler.py [hunk 2]
```diff
@@ -94,3 +101,67 @@ def init_tokenizer(self, model_name="bert-base-cased", kwargs=None):
...
+        # inside a span label of 1
+        # punctuation is ignored in the calculation of metrics: set to -100
+        new_labels = [
+            -100 if word_id is None else labels[word_id] for word_id in word_ids
+        ]
+        # if the beginning of a span has been split into two tokens,
+        # make sure that the label "2" only appears once
+        # seems to me we need to use enumerate
+        new_labels = [
+            1 if label == 2 and new_labels[i - 1] == 2 and i >= 1 else label
```
Comment (lkeegan) [inline moralization/transformers_data_handler.py:137]: I think the check that i>0 should come before indexing `i-1`
```suggestion
            1 if label == 2 and i >= 1 and new_labels[i - 1] == 2 else label
```

File: moralization/transformers_data_handler.py [hunk 3]
```diff
@@ -94,3 +101,67 @@ def init_tokenizer(self, model_name="bert-base-cased", kwargs=None):
...
+    def _check_is_nested(self, list_to_check: list) -> list:
+        # make sure that data is a nested list,
+        # otherwise add a layer
+        # do we need this?
+        # maybe enough to check if list_to_check[0] is a list?
+        # does it cost us to iterate over all the data?
+        list_to_check = (
+            [list_to_check] if not isinstance(list_to_check[0], list) else list_to_check
+        )
+        list_to_check = [[i] if not isinstance(i, list) else i for i in list_to_check]
```
Comment (lkeegan) [inline moralization/transformers_data_handler.py:122]: may be worth checking if the list is already nested first before making a new list? (it means checking twice when the list is not nested, but I would guess making a new list is the expensive part here)
```suggestion
        if not all(isinstance(i, list) for i in list_to_check):
            list_to_check = [[i] if not isinstance(i, list) else i for i in list_to_check]
```

File: moralization/transformers_data_handler.py
Comment (GwydionJon) [inline moralization/transformers_data_handler.py:1]: apart from liams comments this lgtm

File: moralization/tests/test_transformers_data_handler.py [hunk 4]
```diff
@@ -16,6 +17,29 @@ def gen_instance():
     return TransformersDataHandler()
 
 
+@pytest.fixture(scope="module")
+def raw_dataset():
+    return load_dataset("iulusoy/test-data")
```
[the remark this answers is not in the document]
Comment (iulusoy) [reply moralization/tests/test_transformers_data_handler.py:22]: This is the same data as in the repo, just in dataset format - we could add the csv?

**Pull request merged**